docs/en/advanced-concepts/migrating-to-appium-2.0.md

```diff
@@ -127,3 +127,82 @@ TODO
+
+With Appium 2.0, we enter a new era where end users may wish to target various independent versions of drivers and plugins. With Appium 1.x, this was never the case, since any given version of Appium would contain one and only one version of each driver. It is of course up to each service provider how they wish to implement the discovery, installation, and availability of any official or third party drivers or plugins. But the Appium team wants to make a recommendation in terms of the capabilities service providers support, for consistency across the industry. This is a recommendation only, and not a standard, but adopting it will help users to navigate the increased complexity that working with Appium 2.0 in a cloud environment may bring.
+
+### Suggested capabilities
+
+In addition to the standard `platformName`, `appium:deviceName`, `appium:automationName`, and `appium:platformVersion`, we recommend adopting the following set, where the label `$cloud` is not meant to be interpreted literally but instead should be replaced by your vendor prefix (so for HeadSpin it would be `headspin`, Sauce Labs it would be `sauce`, and BrowserStack it would be `browserstack`, to name just a few examples):
+
+|Capability|Used for|Example|
+|----------|-------|-------|
+|`$cloud:appiumVersion`|Designating which version of the Appium server is used to host and manage drivers. If ommitted, behavior left up to the provider, but the recommendation would be to provide the latest official version.|`2.0.0`|
+|`$cloud:automationVersion`|Designating which version of the specified driver should be used.|`1.55.2`|
+|`$cloud:automation`|Designating a custom driver to use (see below for more info). This would override `appium:automationName` and `$cloud:automationVersion`|`{"name": "org/custom-driver", "source": "github", "package": "custom-driver"}`|
+|`$cloud:appiumPlugins`|Designating the list of plugins (and potentially versions of plugins) to be activated (see below for more info).|`["images", "universal-xml"]`|
+
+### Basic example
+
+Appium extensions (drivers and plugins) have a set of properties that specify where they can be installed from. Cloud providers are obviously under no obligation to provide support for arbitrarily specified extensions, seeing as these may represent untrusted code running in a managed environment. In the case where arbitrary extensions are not supported, the `appium:automationName`, `$cloud:automationVersion`, and `$cloud:appiumPlugins` capabilities should be sufficient. See the following JSON object representing capabilities for a session:
+
+```json
+{
+    "$cloud:appiumVersion": "2.0.0",
+    "platformName": "iOS",
+    "appium:platformVersion": "14.4",
+    "appium:deviceName": "iPhone 11",
+    "appium:app": "Some-App.app.zip",
+    "appium:automationName": "XCUITest",
+    "$cloud:automationVersion": "3.52.0",
+    "$cloud:appiumPlugins": ["images"]
+}
```

Comment:
I could imagine this to work better:
```js
{
    "platformName": "iOS",
    "$cloud:appiumOptions": {
        "version": "2.0.0",
        "platformVersion": "14.4",
        "deviceName": "iPhone 11",
        "app": "Some-App.app.zip",
        "automationName": "XCUITest",
        "automationVersion": "3.52.0",
        "plugins": ["images"]
    }
}
```

It is not only consistent with what browserdriver do (e.g. `'goog:chromeOptions'`) , it also seems less repetitive in terms of having appium in the key. Furthermore I don't really see the necessity to differentiate between `appium:automationName` and `$cloud:automationVersion` , why would the one be Appium specific and the other cloud vendor specific?

Reply:
After thinking a bit more I believe I understood now: locally vs running in the cloud. `automationVersion` doesn't really make sense locally as you only have the version that you have installed. I would still tend to the this format given it is less verbose and more aligned to other implementations.

Reply:
Yes, `appium:automationName` is a standard Appium capability that the Appium server understands, and it means the least amount of changes when adjusting caps from local to cloud. IMO clouds should support both `appium:automationName` and `$cloud:automationName`.

Reply:
the idea of putting it all into `$cloud:appiumOptions` is interesting though, let's think about that.

Reply:
On the Sauce Labs side we start to follow exactly this pattern as it better allows to group certain domain settings and also to extend them with new capabilities, e.g. we have `"sauce:visual": { ... }` and soon something like `"sauce:network": { ... }`.

The only downside I see from this grouping is that it makes it more difficult to differentiate between native Appium capabilities and possible cloud vendor capabilities, e.g. here you see the difference:

```js
    "appium:automationName": "XCUITest",
    "$cloud:automationVersion": "3.52.0",
```

but with grouping they are thrown into one bucket:

```js
    "$cloud:appiumOptions": {
        "automationName": "XCUITest",
        "automationVersion": "3.52.0"
    }
```

That said, I am not sure if this can conflict if Appium extends their caps with one that is already being used by a vendor. Then grouping would not work anymore but it would be still possible to differentiate when using proposed suggestion, e.g.:

```js
    "appium:foobar": "...",
    "$cloud:foobar": "...",
```

Another way to look at this is to group cloud specific Appium options and native ones:

```js
{
    "platformName": "iOS",
    "$cloud:appiumOptions": {
        "version": "2.0.0",
        "automationVersion": "3.52.0",
        "plugins": ["images"]
     },
     "appium:options": {
        "platformVersion": "14.4",
        "deviceName": "iPhone 11",
        "app": "Some-App.app.zip",
        "automationName": "XCUITest",
    }
}
```

This even seems much cleaner to me. It allows extending both without conflicts and avoids duplication.

Reply:
Personally. when appium gets group section as its capabilities, I like the below format as suggested example since users can keep appium options as-is between their local and cloud.  (of course, cloud vendors can accept all capabilities as `$cloud:appiumOptions` and convert appium specific capabilities into `appium:` or `appium:options` (when it is available) in their internal for their local appium. It depends on their decision).

```json
{
    "platformName": "iOS",
    "$cloud:appiumOptions": {
        "version": "2.0.0",
        "automationVersion": "3.52.0",
        "plugins": ["images"]
     },
     "appium:options": {
        "platformVersion": "14.4",
        "deviceName": "iPhone 11",
        "app": "Some-App.app.zip",
        "automationName": "XCUITest",
    }
}
```

Current appium clients (in appium repository) adds `appium:` prefix for each capability that has no vendor prefix and non-w3c standard capabilities to reduce end-user side changes. When we support this style, it would be better to put such caps into appium options. (Then, will work for this)

Reply:
I like this idea too. Two things to consider though:

1. If we want to recommend cloud providers support `appium:options`, then we should also support it in the Appium 2.0 server, otherwise it'd mean too much moving around of caps when deciding whether to run on local or remote.
2. Tools like Appium Inspector are not set up to handle objects within the cap structure. There's currently no way for something like `appium:options` to work in the Inspector without manually writing the JSON. So this would entail doing some work there.

But I agree, I like the clean separation of this.

Reply:
took a stab at implementing support for the capability within the appium server: https://github.com/appium/appium/pull/15765

Reply:
👍

Reply:
ok, that's been merged and this doc updated with the new style